// marko/dist/components/Component.js
"use strict";
/* jshint newcap:false */

var complain;

var domInsert = require("../runtime/dom-insert");
var defaultCreateOut = require("../runtime/createOut");
var getComponentsContext = require("./ComponentsContext").__;
var componentsUtil = require("./util");
var componentLookup = componentsUtil.a_;
var emitLifecycleEvent = componentsUtil.b_;
var destroyNodeRecursive = componentsUtil.c_;
var EventEmitter = require("events-light");
var RenderResult = require("../runtime/RenderResult");
var SubscriptionTracker = require("listener-tracker");
var inherit = require("raptor-util/inherit");
var updateManager = require("./update-manager");
var morphdom = require("../morphdom");
var eventDelegation = require("./event-delegation");
var domData = require("./dom-data");
var componentsByDOMNode = domData.d_;
var CONTEXT_KEY = "__subtree_context__";

var slice = Array.prototype.slice;

var COMPONENT_SUBSCRIBE_TO_OPTIONS;
var NON_COMPONENT_SUBSCRIBE_TO_OPTIONS = {
    addDestroyListener: false
};

var emit = EventEmitter.prototype.emit;
var ELEMENT_NODE = 1;

function removeListener(removeEventListenerHandle) {
    removeEventListenerHandle();
}

function handleCustomEventWithMethodListener(component, targetMethodName, args, extraArgs) {
    // Remove the "eventType" argument
    args.push(component);

    if (extraArgs) {
        args = extraArgs.concat(args);
    }

    var targetComponent = componentLookup[component.e_];
    var targetMethod = typeof targetMethodName === "function" ? targetMethodName : targetComponent[targetMethodName];
    if (!targetMethod) {
        throw Error("Method not found: " + targetMethodName);
    }

    targetMethod.apply(targetComponent, args);
}

function resolveKeyHelper(key, index) {
    return index ? key + "_" + index : key;
}

function resolveComponentIdHelper(component, key, index) {
    return component.id + "-" + resolveKeyHelper(key, index);
}

/**
 * This method is used to process "update_<stateName>" handler functions.
 * If all of the modified state properties have a user provided update handler
 * then a rerender will be bypassed and, instead, the DOM will be updated
 * looping over and invoking the custom update handlers.
 * @return {boolean} Returns true if if the DOM was updated. False, otherwise.
 */
function processUpdateHandlers(component, stateChanges, oldState) {
    var handlerMethod;
    var handlers;

    for (var propName in stateChanges) {
        if (stateChanges.hasOwnProperty(propName)) {
            var handlerMethodName = "update_" + propName;

            handlerMethod = component[handlerMethodName];
            if (handlerMethod) {
                (handlers || (handlers = [])).push([propName, handlerMethod]);
            } else {
                // This state change does not have a state handler so return false
                // to force a rerender
                return;
            }
        }
    }

    // If we got here then all of the changed state properties have
    // an update handler or there are no state properties that actually
    // changed.
    if (handlers) {
        // Otherwise, there are handlers for all of the changed properties
        // so apply the updates using those handlers

        handlers.forEach(function (handler) {
            var propertyName = handler[0];
            handlerMethod = handler[1];

            var newValue = stateChanges[propertyName];
            var oldValue = oldState[propertyName];
            handlerMethod.call(component, newValue, oldValue);
        });

        emitLifecycleEvent(component, "update");

        component.f_();
    }

    return true;
}

function checkInputChanged(existingComponent, oldInput, newInput) {
    if (oldInput != newInput) {
        if (oldInput == null || newInput == null) {
            return true;
        }

        var oldKeys = Object.keys(oldInput);
        var newKeys = Object.keys(newInput);
        var len = oldKeys.length;
        if (len !== newKeys.length) {
            return true;
        }

        for (var i = 0; i < len; i++) {
            var key = oldKeys[i];
            if (oldInput[key] !== newInput[key]) {
                return true;
            }
        }
    }

    return false;
}

var componentProto;

/**
 * Base component type.
 *
 * NOTE: Any methods that are prefixed with an underscore should be considered private!
 */
function Component(id) {
    EventEmitter.call(this);
    this.id = id;
    this.g_ = null;
    this.h_ = null;
    this.i_ = null;
    this.j_ = null;
    this.k_ = null; // Used to keep track of bubbling DOM events for components rendered on the server
    this.l_ = null;
    this.e_ = null;
    this.m_ = null;
    this.n_ = undefined;
    this.o_ = false;
    this.p_ = undefined;

    this.q_ = false;
    this.r_ = false;
    this.s_ = false;
    this.t_ = false;

    this.u_ = undefined;

    this.v_ = {};
    this.w_ = undefined;
}

Component.prototype = componentProto = {
    x_: true,

    subscribeTo: function (target) {
        if (!target) {
            throw TypeError();
        }

        var subscriptions = this.i_ || (this.i_ = new SubscriptionTracker());

        var subscribeToOptions = target.x_ ? COMPONENT_SUBSCRIBE_TO_OPTIONS : NON_COMPONENT_SUBSCRIBE_TO_OPTIONS;

        return subscriptions.subscribeTo(target, subscribeToOptions);
    },

    emit: function (eventType) {
        var customEvents = this.l_;
        var target;

        if (customEvents && (target = customEvents[eventType])) {
            var targetMethodName = target[0];
            var isOnce = target[1];
            var extraArgs = target[2];
            var args = slice.call(arguments, 1);

            handleCustomEventWithMethodListener(this, targetMethodName, args, extraArgs);

            if (isOnce) {
                delete customEvents[eventType];
            }
        }

        if (this.listenerCount(eventType)) {
            return emit.apply(this, arguments);
        }
    },
    getElId: function (key, index) {
        return resolveComponentIdHelper(this, key, index);
    },
    getEl: function (key, index) {
        if (key) {
            var resolvedKey = resolveKeyHelper(key, index);
            var keyedElement = this.v_["@" + resolvedKey];

            if (!keyedElement) {
                var keyedComponent = this.getComponent(resolvedKey);

                if (keyedComponent) {
                    return keyedComponent.h_.firstChild;
                    // eslint-disable-next-line no-constant-condition
                }
            }

            return keyedElement;
        } else {
            return this.el;
        }
    },
    getEls: function (key) {
        key = key + "[]";

        var els = [];
        var i = 0;
        var el;
        while (el = this.getEl(key, i)) {
            els.push(el);
            i++;
        }
        return els;
    },
    getComponent: function (key, index) {
        var rootNode = this.v_[resolveKeyHelper(key, index)];
        if (/\[\]$/.test(key)) {
            rootNode = rootNode && rootNode[Object.keys(rootNode)[0]];
            // eslint-disable-next-line no-constant-condition
        }
        return rootNode && componentsByDOMNode.get(rootNode);
    },
    getComponents: function (key) {
        var lookup = this.v_[key + "[]"];
        return lookup ? Object.keys(lookup).map(function (key) {
            return componentsByDOMNode.get(lookup[key]);
        }) : [];
    },
    destroy: function () {
        if (this.q_) {
            return;
        }

        var root = this.h_;
        var nodes = this.h_.nodes;

        this.y_();

        nodes.forEach(function (node) {
            destroyNodeRecursive(node);

            if (eventDelegation.z_(node) !== false) {
                node.parentNode.removeChild(node);
            }
        });

        root.detached = true;

        delete componentLookup[this.id];
        this.v_ = {};
    },

    y_: function () {
        if (this.q_) {
            return;
        }

        emitLifecycleEvent(this, "destroy");
        this.q_ = true;

        componentsByDOMNode.set(this.h_, undefined);

        this.h_ = null;

        // Unsubscribe from all DOM events
        this.A_();

        var subscriptions = this.i_;
        if (subscriptions) {
            subscriptions.removeAllListeners();
            this.i_ = null;
        }
    },

    isDestroyed: function () {
        return this.q_;
    },
    get state() {
        return this.g_;
    },
    set state(newState) {
        var state = this.g_;
        if (!state && !newState) {
            return;
        }

        if (!state) {
            state = this.g_ = new this.B_(this);
        }

        state.C_(newState || {});

        if (state.s_) {
            this.D_();
        }

        if (!newState) {
            this.g_ = null;
        }
    },
    setState: function (name, value) {
        var state = this.g_;

        if (typeof name == "object") {
            // Merge in the new state with the old state
            var newState = name;
            for (var k in newState) {
                if (newState.hasOwnProperty(k)) {
                    state.E_(k, newState[k], true /* ensure:true */);
                }
            }
        } else {
            state.E_(name, value, true /* ensure:true */);
        }
    },

    setStateDirty: function (name, value) {
        var state = this.g_;

        if (arguments.length == 1) {
            value = state[name];
        }

        state.E_(name, value, true /* ensure:true */
        , true /* forceDirty:true */
        );
    },

    replaceState: function (newState) {
        this.g_.C_(newState);
    },

    get input() {
        return this.n_;
    },
    set input(newInput) {
        if (this.t_) {
            this.n_ = newInput;
        } else {
            this.F_(newInput);
        }
    },

    F_: function (newInput, onInput, out) {
        onInput = onInput || this.onInput;
        var updatedInput;

        var oldInput = this.n_;
        this.n_ = undefined;
        this.G_ = out && out[CONTEXT_KEY] || this.G_;

        if (onInput) {
            // We need to set a flag to preview `this.input = foo` inside
            // onInput causing infinite recursion
            this.t_ = true;
            updatedInput = onInput.call(this, newInput || {}, out);
            this.t_ = false;
        }

        newInput = this.m_ = updatedInput || newInput;

        if (this.s_ = checkInputChanged(this, oldInput, newInput)) {
            this.D_();
        }

        if (this.n_ === undefined) {
            this.n_ = newInput;
            if (newInput && newInput.$global) {
                this.p_ = newInput.$global;
            }
        }

        return newInput;
    },

    forceUpdate: function () {
        this.s_ = true;
        this.D_();
    },

    D_: function () {
        if (!this.r_) {
            this.r_ = true;
            updateManager.H_(this);
        }
    },

    update: function () {
        if (this.q_ === true || this.I_ === false) {
            return;
        }

        var input = this.n_;
        var state = this.g_;

        if (this.s_ === false && state !== null && state.s_ === true) {
            if (processUpdateHandlers(this, state.J_, state.K_, state)) {
                state.s_ = false;
            }
        }

        if (this.I_ === true) {
            // The UI component is still dirty after process state handlers
            // then we should rerender

            if (this.shouldUpdate(input, state) !== false) {
                this.L_(false);
            }
        }

        this.f_();
    },

    get I_() {
        return this.s_ === true || this.g_ !== null && this.g_.s_ === true;
    },

    f_: function () {
        this.s_ = false;
        this.r_ = false;
        this.m_ = null;
        var state = this.g_;
        if (state) {
            state.f_();
        }
    },

    shouldUpdate: function () {
        return true;
    },

    b_: function (eventType, eventArg1, eventArg2) {
        emitLifecycleEvent(this, eventType, eventArg1, eventArg2);
    },

    L_: function (isRerenderInBrowser) {
        var self = this;
        var renderer = self.M_;

        if (!renderer) {
            throw TypeError();
        }

        var rootNode = this.h_;

        var doc = self.u_;
        var input = this.m_ || this.n_;
        var globalData = this.p_;

        updateManager.N_(function () {
            var createOut = renderer.createOut || defaultCreateOut;
            var out = createOut(globalData);
            out.sync();
            out.u_ = self.u_;
            out[CONTEXT_KEY] = self.G_;

            var componentsContext = getComponentsContext(out);
            var globalComponentsContext = componentsContext.O_;
            globalComponentsContext.P_ = self;
            globalComponentsContext.Q_ = isRerenderInBrowser;

            renderer(input, out);

            var result = new RenderResult(out);

            var targetNode = out.R_().S_;

            morphdom(rootNode, targetNode, doc, componentsContext);

            result.afterInsert(doc);
        });

        this.f_();
    },

    T_: function () {
        var root = this.h_;
        root.remove();
        return root;
    },

    A_: function () {
        var eventListenerHandles = this.j_;
        if (eventListenerHandles) {
            eventListenerHandles.forEach(removeListener);
            this.j_ = null;
        }
    },

    get U_() {
        var state = this.g_;
        return state && state.V_;
    },

    W_: function (customEvents, scope) {
        var finalCustomEvents = this.l_ = {};
        this.e_ = scope;

        customEvents.forEach(function (customEvent) {
            var eventType = customEvent[0];
            var targetMethodName = customEvent[1];
            var isOnce = customEvent[2];
            var extraArgs = customEvent[3];

            finalCustomEvents[eventType] = [targetMethodName, isOnce, extraArgs];
        });
    },

    get el() {
        return this.h_ && this.h_.firstChild;
        // eslint-disable-next-line no-constant-condition
    },

    get els() {
        return (this.h_ ? this.h_.nodes : []).filter(function (el) {
            return el.nodeType === ELEMENT_NODE;
        });
        // eslint-disable-next-line no-constant-condition
    }
};

componentProto.elId = componentProto.getElId;
componentProto.X_ = componentProto.update;
componentProto.Y_ = componentProto.destroy;

// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(componentProto, function getEl(component) {
    return component.T_();
}, function afterInsert(component) {
    return component;
});

inherit(Component, EventEmitter);

module.exports = Component;
